substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java

```diff
@@ -0,0 +1,134 @@
+            if (targetConstructor != null) {
+                serializationSupport.addSerializationConstructorAccessorClass(serializationTargetClass, parameterTypes, checkedExceptions, modifiers, targetConstructor);
+            }
+            addReflections(reflectionData, serializationTargetClass, targetConstructor);
+        };
+
+        SerializationConfigurationParser parser = new SerializationConfigurationParser(serializationAdapter);
+        loadedConfigurations = ConfigurationParserUtils.parseAndRegisterConfigurations(parser, imageClassLoader, "serialization",
+                        ConfigurationFiles.Options.SerializationConfigurationFiles, ConfigurationFiles.Options.SerializationConfigurationResources,
+                        ConfigurationFiles.SERIALIZATION_NAME);
+        String exceptionsMsg = serializationSupport.collectMultiDefinitions();
+        // Checkstyle: stop
+        if (exceptionsMsg.length() > 0) {
+            System.out.println(exceptionsMsg);
+            if (!NativeImageOptions.ReportUnsupportedElementsAtRuntime.getValue()) {
+                BigBang bb = access.getBigBang();
+                bb.getUnsupportedFeatures().addMessage("Unsupported dynamic features", null,
+                                "To allow continuing compilation with above unsupported features, set " +
+                                                SubstrateOptionsParser.commandArgument(NativeImageOptions.ReportUnsupportedElementsAtRuntime, "+"));
+            } else {
+                System.out.println("Compilation will continue because " + SubstrateOptionsParser.commandArgument(NativeImageOptions.ReportUnsupportedElementsAtRuntime, "+") +
+                                " was set. But the program may behave unexpectedly at runtime.");
+            }
+        }
+        // Checkstyle: resume
+    }
+
+    public static void addReflections(ReflectionDataBuilder reflectionData, Class<?> serializationTargetClass, Class<?> targetConstructor) {
+        if (targetConstructor != null) {
+            reflectionData.register(targetConstructor.getDeclaredConstructors());
+        }
+        reflectionData.register(serializationTargetClass);
+        for (Field f : serializationTargetClass.getDeclaredFields()) {
+            int modifiers = f.getModifiers();
+            boolean allowWrite = false;
+            boolean allowUnsafeAccess = false;
+            if (!(Modifier.isFinal(modifiers) && Modifier.isStatic(modifiers))) {
+                allowWrite = Modifier.isFinal(f.getModifiers());
+                allowUnsafeAccess = !Modifier.isStatic(f.getModifiers());
+            }
+            reflectionData.register(allowWrite, allowUnsafeAccess, f);
+        }
+        reflectionData.register(serializationTargetClass.getDeclaredMethods());
+        reflectionData.register(serializationTargetClass.getDeclaredConstructors());
+    }
```

Comment:
Do all fields, constructors and methods really need to be registered? It seems to me that only those that are singled out in the `ObjectStreamClass(Class<?>)` constructor are really relevant. But that might be irrelevant with the approach of populating a collection of `ObjectStreamClass` at build time.

Reply:
`ObjectStreamClass.computeDefaultSUID` calls `getDeclaredMethods`, `getDeclaredFields `and `getDeclaredConstructors` to calculate serializationUID when it's absent. So it can be conditionally register all fields, methods and constructors.

`ObjectStreamClass.getDefaultSerialFields` calls `getDeclaredFields` to collect all fields that can be serialized. But the same logic can be applied here to avoid registering all fields.

Reply:
Thanks for clarifying!